deploy-service/common/src/main/java/com/pinterest/deployservice/handler/PingHandler.java

```diff
@@ -363,23 +368,109 @@ public PingResponseBean ping(PingRequestBean pingRequest) throws Exception {
         if (!installCandidates.isEmpty()) {
             GoalAnalyst.InstallCandidate installCandidate = installCandidates.get(0);
             AgentBean updateBean = installCandidate.updateBean;
-            if (installCandidate.needWait) {
-                LOG.debug("Checking if host {}, updateBean = {} can deploy", hostName, updateBean);
-                if (canDeploy(installCandidate.env, hostName, updateBean)) {
-                    // use the updateBean in the installCandidate instead
-                    LOG.debug("Host {} can proceed to deploy, updateBean = {}", hostName, updateBean);
+            EnvironBean env = installCandidate.env;
+            String scheduleId = env.getSchedule_id();
+
+            if (scheduleId == null) {
+                if (installCandidate.needWait) { // new candidate 
+                    LOG.debug("Checking if host {}, updateBean = {} can deploy", hostName, updateBean);
+                    if (canDeploy(env, hostName, updateBean)) {
+                        LOG.debug("Host {} can proceed to deploy, updateBean = {}", hostName, updateBean);
+                        updateBeans.put(updateBean.getEnv_id(), updateBean);
+                        response = generateInstallResponse(installCandidate);
+                    } else {
+                        LOG.debug("Host {} for env {} needs to wait for its turn to install.",
+                            hostName, updateBean.getEnv_id());
+                    }
+                } else {
+                    LOG.debug("Host {} is in the middle of deploy, no need to wait, updateBean = {}",
+                        hostName, updateBean);
                     updateBeans.put(updateBean.getEnv_id(), updateBean);
                     response = generateInstallResponse(installCandidate);
-                } else {
-                    LOG.debug("Host {} for env {} needs to wait for its turn to install.",
-                            hostName, updateBean.getEnv_id());
                 }
             } else {
-                LOG.debug("Host {} is in the middle of deploy, no need to wait, updateBean = {}",
+                ScheduleBean schedule = scheduleDAO.getById(scheduleId);
```

Comment:
do you mean the entire new logic? A lot if it is dependent on the code before such as variables and such though